python/cuspatial/cuspatial/geometry/pygeoarrow.py

```diff
@@ -0,0 +1,189 @@
+# Copyright (c) 2021 NVIDIA CORPORATION
+
+import geopandas as gpd
+import pyarrow as pa
+from shapely.geometry import (
+    LineString,
+    MultiLineString,
+    MultiPoint,
+    MultiPolygon,
+    Point,
+    Polygon,
+)
+
+
+def named_list(name, x, size=-1):
+    return pa.list_(pa.field(name, x, nullable=False), list_size=size)
+
+
+XYType: pa.ListType = named_list("xy", pa.float64(), size=2)
+
+ArrowPolygonsType: pa.ListType = named_list(
+    "polygons", named_list("rings", named_list("vertices", XYType))
+)
+
+ArrowLinestringsType: pa.ListType = named_list(
+    "lines", named_list("offsets", XYType)
+)
+
+ArrowMultiPointsType: pa.ListType = named_list("points", XYType)
+
+ArrowPointsType: pa.ListType = XYType
+
+
+def getGeoArrowUnionRootType() -> pa.union:
+    return pa.union(
+        [
+            ArrowPointsType,
+            ArrowMultiPointsType,
+            ArrowLinestringsType,
+            ArrowPolygonsType,
+        ],
+        mode="dense",
+    )
+
+
+def from_geopandas(geoseries: gpd.GeoSeries) -> pa.lib.UnionArray:
+    def get_coordinates(data) -> tuple:
+        point_coords = []
+        mpoint_coords = []
+        line_coords = []
+        polygon_coords = []
+        all_offsets = [0]
+        type_buffer = []
+
+        for geom in data:
+            coords = geom.__geo_interface__["coordinates"]
+            if isinstance(geom, Point):
+                point_coords.append(coords)
+                all_offsets.append(1)
+            elif isinstance(geom, MultiPoint):
+                mpoint_coords.append(coords)
+                all_offsets.append(len(coords))
+            elif isinstance(geom, LineString):
+                line_coords.append([coords])
+                all_offsets.append(1)
+            elif isinstance(geom, MultiLineString):
+                line_coords.append(coords)
+                all_offsets.append(len(coords))
+            elif isinstance(geom, Polygon):
+                polygon_coords.append([coords])
+                all_offsets.append(1)
+            elif isinstance(geom, MultiPolygon):
+                polygon_coords.append(coords)
+                all_offsets.append(len(coords))
+            else:
+                raise TypeError(type(geom))
+            type_buffer.append(
+                {
+                    Point: 0,
+                    MultiPoint: 1,
+                    LineString: 2,
+                    MultiLineString: 3,
+                    Polygon: 4,
+                    MultiPolygon: 5,
+                }[type(geom)]
```

Comment:
@thomcom Sorry I'm late to the review game here. Isn't it slow to create a new dict and do this lookup for every value in the `GeoSeries`? Why not append the typecode in the conditionals above (like we do with the offsets), or at least define this dict at the top once and read from it?

Reply:
Thanks for the comment, I'll clean that up on the next go around, which is coming up today.

Reply:
I just moved it into the append step like you suggested. 3 appends seems like a lot. :parrot: